ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -530,12 +541,45 @@ RouterException getChunkException() {
     void poll(RequestRegistrationCallback<GetOperation> requestRegistrationCallback) {
       //First, check if any of the existing requests have timed out.
       cleanupExpiredInFlightRequests();
-      checkAndMaybeComplete();
-      if (!isComplete()) {
-        fetchRequests(requestRegistrationCallback);
+      if (!maybeProcessCallbacksAndCompleteOperation()) {
+        checkAndMaybeComplete();
+        if (!isComplete()) {
+          fetchRequests(requestRegistrationCallback);
+        }
       }
     }
 
+    /**
+     * Maybe process callbacks and complete the operation if applicable. This is a no-op for blobs that doesn't need
+     * any async processing. As of now, decryption is the only async processing that could happen if applicable.
+     * @return {@code true} if the operation is complete, {@code false} otherwise
+     */
+    protected boolean maybeProcessCallbacksAndCompleteOperation() {
+      if (progressTracker.isDecryptionRequired() && decryptCallbackResultInfo.decryptJobComplete) {
+        logger.trace("Processing decrypt callback stored result for data chunk {}", chunkBlobId);
+        if (decryptCallbackResultInfo.exception == null && (getChunkException() == null
```

Comment:
Is the second check really necessary? Will this method get called even that condition is met?

Reply:
I don't think so. Will remove it.